irrd/updates/suspension.py

```diff
@@ -0,0 +1,123 @@
+import logging
+from typing import List, Set, Dict, Tuple
+import functools
+
+from irrd.storage.models import JournalEntryOrigin
+from irrd.conf import config_init, get_setting
+from irrd.rpsl.rpsl_objects import RPSLMntner, rpsl_object_from_text
+from irrd.storage.database_handler import DatabaseHandler
+from irrd.storage.queries import RPSLDatabaseQuery, RPSLDatabaseSuspendedQuery
+from irrd.rpsl.parser import RPSLObject
+from irrd.scopefilter.validators import ScopeFilterValidator
+from irrd.rpki.validators import SingleRouteROAValidator
+
+logger = logging.getLogger(__name__)
+
+
+def suspend_for_mntner(database_handler: DatabaseHandler, suspended_mntner: RPSLMntner) -> List[Dict[str, str]]:
+    source = suspended_mntner.source()
+    if not get_setting(f'sources.{source}.authoritative'):
+        raise ValueError(f'Not authoritative for source {source}')
+
+    logger.info(f"{suspended_mntner.pk()}: Starting suspension for {suspended_mntner}")
+
+    @functools.lru_cache(maxsize=50)
+    def mntner_active(rpsl_pk: str):
+        q = RPSLDatabaseQuery(column_names=['pk']).sources([source]).rpsl_pk(rpsl_pk).object_classes(['mntner'])
+        return bool(list(database_handler.execute_query(q.first_only())))
+
+    suspended_mntner_rpsl_pk = suspended_mntner.pk()
+    # This runs two queries, to account for the suspension of a mntner
+    # who is not an mnt-by for itself. In that case, query1 will not retrieve it,
+    # but query2 will.
+    query1 = RPSLDatabaseQuery(column_names=['pk', 'rpsl_pk', 'object_class', 'source', 'parsed_data'])
+    query1 = query1.sources([source]).lookup_attr('mnt-by', suspended_mntner_rpsl_pk)
+    query1_result = list(database_handler.execute_query(query1))
+    query2 = RPSLDatabaseQuery(column_names=['pk', 'rpsl_pk', 'object_class', 'source', 'parsed_data'])
+    query2 = query2.sources([source]).rpsl_pk(suspended_mntner_rpsl_pk).object_classes(['mntner'])
+    query2_result = list(database_handler.execute_query(query2))
+
+    if not query2_result:
+        msg = f"mntner {suspended_mntner.pk()} does not exist in {source} (or may already be suspended)"
+        logger.info(f"{suspended_mntner.pk()}: error: {msg}")
+        raise ValueError(msg)
+
+    suspended_objects = []
+
+    for row in query1_result + query2_result:
+        if row in suspended_objects:
+            continue
+
+        mntners: Set[str] = set(row['parsed_data']['mnt-by'])
+        mntners.remove(suspended_mntner_rpsl_pk)
```

Comment:
maybe this should be folded into one comprehension, this line is a bit confusing